File: client/src/app/food/model/meal-production-record.ts
import {FoodItem} from './food-item';
import {MealEvent} from './meal-event';

export interface MealProductionFoodItem {
  id: string,
  mpr?: MealProductionRecord,
  foodItem: FoodItem,
  required: number,
  calcRequired?: number,
  prepared: number,
  uom: string
}

export interface MealAttendanceRecord {
  id: string,
  mpr?: MealProductionRecord,
  ageGroup: string,
  projected: number,
  actual: number
}

export interface MealProductionRecord {
  id: string,
  mealDate: Date,
  type: string,
  locked?: boolean,
  lockDate?: Date,
  mealEvent?: MealEvent,
  notes: string,
  attendanceRecords?: MealAttendanceRecord[],
  productionFoodItems?: MealProductionFoodItem[]
}

File: server/src/main/java/com/remarkablesystems/childwatch/domain/food/mpr/MealProductionRecord.java
package com.remarkablesystems.childwatch.domain.food.mpr;

import java.util.Date;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

import javax.persistence.CollectionTable;
import javax.persistence.Column;
import javax.persistence.ElementCollection;
import javax.persistence.Entity;
import javax.persistence.EnumType;
import javax.persistence.Enumerated;
import javax.persistence.Id;
import javax.persistence.JoinColumn;
import javax.persistence.ManyToOne;
import javax.persistence.MapKey;
import javax.persistence.MapKeyJoinColumn;
import javax.persistence.OneToMany;
import javax.persistence.Table;
import javax.persistence.Temporal;
import javax.persistence.TemporalType;

import com.remarkablesystems.childwatch.domain.AuditedUser;
import com.remarkablesystems.childwatch.domain.food.AgeGroup;
import com.remarkablesystems.childwatch.domain.food.Meal;
import com.remarkablesystems.childwatch.domain.food.MealType;

import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.Setter;
import lombok.ToString;

@Entity
@Table(name="meal_production_record")
@ToString(callSuper=true)
@EqualsAndHashCode(callSuper=true)
public class MealProductionRecord extends AuditedUser {

	@Id
	@Getter
	String id;
		
	@Column(name="meal_date")
	@Getter
	@Temporal( TemporalType.DATE )
	Date mealDate;
	
	@Enumerated(EnumType.STRING)
	@Column(name="meal_type")
	@Getter @Setter	
	private MealType type;
	
	@Column(name="locked")
	@Getter @Setter
	boolean locked;
	
	@Column(name="lock_date")
	@Getter @Setter
	@Temporal( TemporalType.DATE )
	Date lockDate;
	
	@ManyToOne
	@JoinColumn(name="meal_id")
	@Getter
	Meal meal;

	@Column(name="notes")
	@Getter @Setter
	String notes;
		
}
